src/dali/abstract_ccxt_input_plugin.py

```diff
@@ -0,0 +1,900 @@
+        self._process_withdrawals(intra_transactions)
+
+        result.extend(in_transactions)
+        result.extend(out_transactions)
+        result.extend(intra_transactions)
+
+        return result
+
+    ### Multiple Transaction Processing or Pagination Methods
+
+    def _process_deposits(
+        self,
+        intra_transactions: List[IntraTransaction],
+    ) -> None:
+
+        pagination_detail_set: AbstractPaginationDetailSet = self.get_process_deposits_pagination_detail_set()
+        pagination_detail_iterator: AbstractPaginationDetailsIterator = iter(pagination_detail_set)
+        try:
+            while True:
+                pagination_details: _PaginationDetails = next(pagination_detail_iterator)
+                deposits = self.__client.fetch_deposits(
+                    code=pagination_details.symbol,
+                    since=pagination_details.since,
+                    limit=pagination_details.limit,
+                    params=pagination_details.params,
+                )
+                # CCXT returns a standardized response from fetch_deposits. 'info' is the exchange-specific information
+                # in this case from Binance.com
+
+                # {
+                #   'info': {
+                #       'amount': '0.00999800',
+                #       'coin': 'PAXG',
+                #       'network': 'ETH',
+                #       'status': '1',
+                #       'address': '0x788cabe9236ce061e5a892e1a59395a81fc8d62c',
+                #       'addressTag': '',
+                #       'txId': '0xaad4654a3234aa6118af9b4b335f5ae81c360b2394721c019b5d1e75328b09f3',
+                #       'insertTime': '1599621997000',
+                #       'transferType': '0',
+                #       'confirmTimes': '12/12',
+                #       'unlockConfirm': '12/12',
+                #       'walletType': '0'
+                #   },
+                #   'id': None,
+                #   'txid': '0xaad4654a3234aa6118af9b4b335f5ae81c360b2394721c019b5d1e75328b09f3',
+                #   'timestamp': 1599621997000,
+                #   'datetime': '2020-09-09T03:26:37.000Z',
+                #   'network': 'ETH',
+                #   'address': '0x788cabe9236ce061e5a892e1a59395a81fc8d62c',
+                #   'addressTo': '0x788cabe9236ce061e5a892e1a59395a81fc8d62c',
+                #   'addressFrom': None,
+                #   'tag': None,
+                #   'tagTo': None,
+                #   'tagFrom': None,
+                #   'type': 'deposit',
+                #   'amount': 0.00999800,
+                #   'currency': 'PAXG',
+                #   'status': 'ok',
+                #   'updated': None,
+                #   'internal': False,
+                #   'fee': None
+                # }
+
+                pagination_detail_iterator.update_fetched_elements(deposits)
+
+                with ThreadPool(self.__thread_count) as pool:
```

Comment:
> I realized that this approach won't really get much speedup via multithreading, because it doesn't multithread the I/O-bound part of the code, which in this case would be fetch_deposits. Is there a way to call fetch_deposits on one single market? That way we could:

You can retrieve all deposits for a code (markets are for trades), but the efficiency of that approach would largely depend on the exchange. Most users wouldn't see a speed increase. Let's take Binance.com for example:

Let's say the user has 100 deposits in 10 different assets just for example's sake. 

Current approach:
1) Pull all deposits - Retrieve all 100 transactions at once. (1 point toward the rate limit)

Proposed approach:
1) fetch available codes (10 points)
2) Make 350+ calls, one for each of the possible assets held at Binance.com. Most of these will return nothing, but will still have to be checked. (350+ x 1 points)

I can see a speed advantage if a user runs an arbitrage bot that initiates many deposit/withdrawals with several different assets, but not with the average user. Binance.com actually allows a user to retrieve 1000 records in one call. I think that would cover most users in one call.

I think it will really depend on how the API on the exchange is structured. I'm afraid we don't have enough experience with the 100+ exchanges on CCXT between us to know the best approach. We might need to build what we can now and optimize later. 

I honestly don't see how multithreading could help withdrawals and deposits since they are relatively fast at the moment.  Trades, on Binance.com at least could theoretically be sped up with multithreading, but I'm already hitting the limit so multithreading would most likely not yield any speed improvements. 

Another thing to think about is that on some exchanges, CCXT may actually be pulling all the transactions and then sorting them by market/symbol and returning that, so multithreading would actually be much slower.

The only area that may have some small advantage is to pull trades, deposits, withdrawals, gains, and implicit API in separate threads. Or more likely, pull trades in one thread and everything else in another. This would speed things up for binance.com at least.

However, honestly, we really just don't know enough about the different structures to make a decision at this point. A lot of exchanges allow you to pull all transactions at once (like Coinbase), but others require you to cycle through markets/symbols. There are simply too many factors to really try to optimize right now. Can we just decide on a simple solution for now and move on? We may hash something out, only to have to rework the entire thing once we have more experience and realize something else might be much more efficient.

Reply:
That's fair: we can optimize as we get more experience. I have to admit my thinking on this was informed by my experience with the Coinbase and Coinbase Pro plugins (where multithreading improved performance 2-3x), but CCXT works differently and the example you gave is compelling. Let's revisit once we support more exchanges.